### PR TITLE
4225: Keycloak OIDC integration

### PR DESCRIPTION
**Pull request by collinpreston**
#### Pre-Flight checklist
- [x] Testing
  - [x] Changes have been manually tested
- [x] Settings
  - [x] New settings are documented and present in `app.json`

#### What are the relevant tickets?
https://github.com/mitodl/open-discussions/issues/4225

#### What's this PR do?
Adds OpenID Connect functionality to Open Discussions which allows for authenticating users via Keycloak.

#### How should this be manually tested?
You will need a Keycloak account configured for https://sso-qa.odl.mit.edu/realms/olapps .  You should be able to create your account by visiting https://sso-qa.odl.mit.edu/admin/olapps/console/#/ .

1. Add the following to your .env file
```
SOCIAL_AUTH_OL_OIDC_OIDC_ENDPOINT=https://sso-qa.odl.mit.edu/realms/olapps
OIDC_ENDPOINT=https://sso-qa.odl.mit.edu/realms/olapps
SOCIAL_AUTH_OL_OIDC_KEY=ol-open-discussions-local
SOCIAL_AUTH_OL_OIDC_SECRET=<REQUEST_FROM_PR_AUTHOR>
AUTHORIZATION_URL=https://sso-qa.odl.mit.edu/realms/olapps/protocol/openid-connect/auth
ACCESS_TOKEN_URL=https://sso-qa.odl.mit.edu/realms/olapps/protocol/openid-connect/token
USERINFO_URL=https://sso-qa.odl.mit.edu/realms/olapps/protocol/openid-connect/userinfo
```
2. Using this PR's branch, visit: http://od.odl.local:8063/login/ol-oidc/
3. You should be redirected to a login page. Enter your Keycloak account credentials.
4. If your account credentials were correct, you should be redirected to the Open Discussions index page.

This can also be tested with the following scenario: 

1. You have an Open Discussions user record with the email apple123@fake.edu.  
2. You then visit http://od.odl.local:8063/login/ol-oidc/ and register a new Keycloak user with apple123@fake.edu.  
3. After confirming the email, you should be able to visit http://od.odl.local:8063/ and be logged in with an active session for apple123@fake.edu.  
4. If this user has Django admin in Open Discussion, you should also be able to access the Django Admin console.  
5. You should see two entries for the apple123@fake.edu user at http://od.odl.local:8063/admin/social_django/usersocialauth/ - one with a provider of "email" and the other with a provider of "ol-oidc".

## NOTE

1. If a user has a Keycloak account but no Open Discussions user record, they will be assigned a random username in Open Discussions.

### CODE REVIEW

File: open_discussions/settings.py
Comment (rhysyngsun) [inline open_discussions/settings.py:388]: I think we should leave these in place for now, is there a reason you were removing them?

**Review by rhysyngsun — CHANGES_REQUESTED** [review]
Code review

File: open_discussions/settings.py
Comment (collinpreston) [reply open_discussions/settings.py:388]: Oh shoot, nope that was me messing around - thank you

**Review by rhysyngsun — APPROVED** [review]
LGTM, we should get the configuration for this in place before merging it though, can you open an issue for that?

**Comment by collinpreston** [comment]
Blocked by https://github.com/mitodl/ol-infrastructure/issues/1775.

**Comment by collinpreston** [comment]
All required environment variables have been added to the Heroku Open Discussion apps in RC and CI.

**Pull request merged**
